### PR TITLE
Fix S3 file's seek operation and retry when connection reset by peer

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: ref #6827

Problem Summary:
After using `PocoHTTPClient`:
1. `S3RandomAccessFile::seek` is failed.
2. Occasionally a connection reset by peer occurs.

### What is changed and how it works?

1. Use `ignore` to replace `seek`.
2. Retry when connection reset by peer occurs.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
  - Run chbenchmark in AWS EC2 and S3.
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Common/ProfileEvents.cpp [hunk 1]
```diff
@@ -137,6 +137,7 @@
     M(S3ListObjects)                           \
     M(S3DeleteObject)                          \
     M(S3CopyObject)                            \
+    M(S3GetObjectRerty)                        \
```
Comment (JaySon-Huang) [inline dbms/src/Common/ProfileEvents.cpp:140]:
```suggestion
    M(S3GetObjectRetry)                        \
```

File: dbms/src/Storages/S3/PocoHTTPClient.cpp [hunk 2]
```diff
@@ -306,21 +304,21 @@ void PocoHTTPClient::makeRequestInternal(
...
             }
             else
             {
                 /// Error statuses are more important so we show them even if `enable_s3_requests_logging == false`.
-                LOG_INFO(log, "Response status: {}, {}", status_code, poco_response.getReason());
+                LOG_INFO(log, "URI={}, Response status: {}, {}", uri, status_code, poco_response.getReason());
             }
 
-            if (poco_response.getStatus() == Poco::Net::HTTPResponse::HTTP_TEMPORARY_REDIRECT)
+            if (poco_response.getStatus() == Poco::Net::HTTPResponse::HTTP_TEMPORARY_REDIRECT || status_code == 302)
```
Comment (JaySon-Huang) [inline dbms/src/Storages/S3/PocoHTTPClient.cpp:315]:
```suggestion
            if (poco_response.getStatus() == Poco::Net::HTTPResponse::HTTP_TEMPORARY_REDIRECT || status_code == Poco::Net::HTTPResponse::HTTP_FOUND)
```

File: dbms/src/Storages/S3/S3RandomAccessFile.cpp [hunk 3]
```diff
@@ -25,79 +25,151 @@
...
 
+#include <chrono>
 #include <optional>
+#include <thread>
 
 namespace ProfileEvents
 {
 extern const Event S3GetObject;
 extern const Event S3ReadBytes;
+extern const Event S3GetObjectRerty;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/S3/S3RandomAccessFile.cpp:36]:
```suggestion
extern const Event S3GetObjectRetry;
```

**Review by JaySon-Huang — APPROVED** [review]
LGTM

File: dbms/src/Storages/S3/S3RandomAccessFile.cpp [hunk 4]
```diff
@@ -25,79 +25,151 @@
...
 
 off_t S3RandomAccessFile::seek(off_t offset_, int whence)
 {
-    if (unlikely(whence != SEEK_SET))
+    while (true)
     {
-        LOG_ERROR(log, "Only SEEK_SET mode is allowed, but {} is received", whence);
-        return -1;
+        auto off = seekImpl(offset_, whence);
+        if (unlikely(off < 0 && errno == 104 && initialize())) // Connection reset by peer and re-initialize succ.
```
Comment (flowbehappy) [inline dbms/src/Storages/S3/S3RandomAccessFile.cpp:92]: Make 104 a constant statict var.

File: dbms/src/Storages/S3/S3RandomAccessFile.cpp [hunk 5]
```diff
@@ -25,79 +25,151 @@
...
     {
-        LOG_ERROR(log, "Seek position is out of bounds. Offset: {}", offset_);
-        return -1;
+        return cur_offset;
     }
     auto & istr = read_result.GetBody();
-    istr.seekg(offset_);
-    return istr.tellg();
+    istr.ignore(offset_ - cur_offset);
+    if (!istr)
```
Comment (flowbehappy) [inline dbms/src/Storages/S3/S3RandomAccessFile.cpp:116]: Should do `if (!istr)` first?

File: dbms/src/Storages/S3/S3RandomAccessFile.h [hunk 6]
```diff
@@ -106,12 +106,18 @@ class S3RandomAccessFile final : public RandomAccessFile
 
     std::shared_ptr<TiFlashS3Client> client_ptr;
     String remote_fname;
-    std::optional<std::pair<UInt64, UInt64>> offset_and_size;
-
+    // We may merge several small files into a S3 object.
+    // `offset_and_size_in_object` is the offset and size of current file in S3 Object.
```
Comment (flowbehappy) [inline dbms/src/Storages/S3/S3RandomAccessFile.h:110]:
```suggestion
    // `offset_and_size_in_object` is the offset and size of current file in S3 Object.
    // And `cur_offset` is the offset in the current file.
```

**Review by flowbehappy — APPROVED** [review]

**Pull request merged**